Improve the dhclient hook for OCI compat

3XX0 opened the pull request:
Now it becomes:
```
lxc.hook.start-host = /usr/share/lxc/hooks/dhclient
lxc.hook.stop = /usr/share/lxc/hooks/dhclient
```

@stgraber @brauner Do you have any suggestion to address the LSM problems?
Most distribution protect `dhclient`, thus it won't have permissions to read or write config/lease/pid files. Should we write Apparmor/SELinux profile overrides or let the users figure it out (e.g. document what to do in `/etc/apparmor.d/local/sbin.dhclient`)?

@hallyn Seems like the hook namespace fds got fixed so I was able to leverage it.
Right now I use `oci.common.conf` to setup the hooks for all my OCI containers, should we add a `--dhcp` option to the OCI template as well? It might be problematic if both are set though, what do you think?

lxc-jenkins commented:
This pull request didn't trigger Jenkins as its author isn't in the whitelist.

An organization member must perform one of the following:

 - To have this branch tested by Jenkins, use the "ok to test" command.
 - To have a one time test done, use the "test this please" command.

Those commands are simple Github comments of the format: "jenkins: COMMAND"

flx42 commented on hooks/dhclient.in:
```diff
@@ -0,0 +1,68 @@
+    rm -f "${pidfile}"
+}
+
+if [ "$#" -lt 3 ] || [ "$2" != "lxc" ]; then
```
I'm using the same test in my script, I would be interested to know if you think that makes sense and if it's safe to have (e.g.: are we expecting new arguments to be passed to hooks in the near-future?)

flx42 commented on templates/lxc-oci.in:
```diff
@@ -279,6 +279,10 @@ if [ -n "$LXC_MAPPED_UID" ] && [ "$LXC_MAPPED_UID" != "-1" ] && [ -e "${LXC_TEMP
 fi
 
+if [ -e "${LXC_TEMPLATE_CONFIG}/oci.common.conf" ]; then
+    echo "lxc.include = ${LXC_TEMPLATE_CONFIG}/oci.common.conf" >> "${LXC_CONF_FILE}"
```
Should we push `oci.common.conf.in` in the same commit? Even if it only contains commented hooks for now, e.g.:
```
# Uncomment if you want to use dhclient...
#lxc.hook.start-host = @LXCHOOKDIR@/dhclient-start
#lxc.hook.stop = @LXCHOOKDIR@/dhclient-stop
```

hallyn commented:
Regarding the hook args, quite the opposite - we are hoping to stop sending them as argv and just send them as environment variables.

hallyn commented:
--dhcp option would be nice.   Can it just double-check whether the dhcp hook is already set in the config when the template is called to avoid double-setting it?

3XX0 commented:
I updated the PR, I added the `--dhcp` option and guarded against multiple inclusion by checking the PID file and I included an `oci.common.conf.in` commented out.

@hallyn @brauner I will update the hook arguments once we settled on the environment variables

brauner commented:
See #2026 .

3XX0 commented:
I updated the hook to use the new environment variables

3XX0 commented:
@brauner I updated the hook to work with your PR

brauner commented:
@3XX0, excellent. Thanks! I'll wait for @hallyn to merge my pr and then I'm merging this one.

brauner commented:
Labeling this as "blocked" until #2026 is merged. :)

brauner commented on hooks/dhclient.in:
```diff
@@ -0,0 +1,91 @@
+        echo "WARN: DHCP client is already running, skipping start hook." >> "${debugfile}"
+    else
+        echo "INFO: Starting DHCP client and acquiring a lease..." >> "${debugfile}"
+        nsenter --uts --user --net --target "${LXC_PID}" -- \
```
So I think this won't work unconditionally. The problem is that with `--user` the host and container need to be in different user namespaces. If they are not this will fail with `EINVAL`. I think what you want to do is look whether `LXC_USER_NS` is set if `LXC_HOOK_VERSION` is set to `1` or if `user:/proc/<pid>/fd/<nr>` is passed as argument when `LXC_HOOK_VERSION` is set to `0` or not set at all.

3XX0 replied:
Right, good catch. Shouldn't we do that for every namespace though since the user could have specified a `--share-<ns>`?

brauner replied:
If I remember correctly, then only user namespaces have this restriction. The other ones should be fine. Was that your concern or did I misunderstand?
(Fwiw, since I introduced `lxc.namespace.[namespace identifier]` users can actually even specify this in their config.)

3XX0 replied:
It works for the `stop` hook but what about the `start-host` hook? Do I have to check the userns proc inodes or there is a better way to know whether it's shared or not?

brauner replied:
One way to do it would be to compare:

    readlink /proc/${LXD_PID}/ns/user == readlink /proc/self/ns/user

brauner replied:
Since by the time the `start-hook` runs we already know the init pid of the container and we make it available in `LXC_PID`.

brauner replied:
Oh, and the `LXC_PID` is guaranteed to be valid until after the `start-host` hook is run since we place a barrier right after it. Only after the `start-hook`s are run is the the child allowed to `exec*()`.

brauner commented on hooks/dhclient.in:
```diff
@@ -0,0 +1,91 @@
+        done
+    else
+        ns_args+=("--uts=${LXC_UTS_NS}")
+        ns_args+=("--user=${LXC_USER_NS}")
```
Same comment applies here: So I think this won't work unconditionally. The problem is that with `--user` the host and container need to be in different user namespaces. If they are not this will fail with `EINVAL`. I think what you want to do is look whether `LXC_USER_NS` is set if `LXC_HOOK_VERSION` is set to `1` or if `user:/proc/<pid>/fd/<nr>` is passed as argument when `LXC_HOOK_VERSION` is set to `0` or not set at all.

3XX0 commented:
- Rebased on master
- Fixed the issue w.r.t the userns
- Ran shellcheck
- Tested  unprivileged/privileged containers with `hook.version` 1 and 2

brauner commented:
jenkins: test this please
